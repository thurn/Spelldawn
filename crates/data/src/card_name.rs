// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use convert_case::{Case, Casing};
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use strum_macros::Display;

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Display, Serialize, Deserialize)]
pub enum CardName {
    /// Empty identity cards with no associated rules text, for use in tests
    TestChampionIdentity,
    TestOverlordIdentity,

    ArcaneRecovery,
    Greataxe,
    GoldMine,
    IceDragon,
    DungeonAnnex,
}

impl CardName {
    pub fn displayed_name(&self) -> String {
        format!("{}", self).from_case(Case::Pascal).to_case(Case::Title)
    }
}

impl PartialOrd<Self> for CardName {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for CardName {
    fn cmp(&self, other: &Self) -> Ordering {
        self.to_string().cmp(&other.to_string())
    }
}
