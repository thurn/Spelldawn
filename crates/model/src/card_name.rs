// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use strum_macros::Display;

use convert_case::{Case, Casing};

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Display)]
pub enum CardName {
    /// Empty identity cards with no associated rules text, for use in unit tests
    TestChampionIdentity,
    TestOverlordIdentity,

    ArcaneRecovery,
    Greataxe,
    GoldMine,
    IceDragon,
    DungeonAnnex,
}

impl CardName {
    pub fn displayed_name(&self) -> String {
        format!("{}", self).from_case(Case::Pascal).to_case(Case::Title)
    }
}
