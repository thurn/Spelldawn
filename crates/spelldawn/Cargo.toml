[package]
name = "spelldawn"
version = "0.0.0"
authors = ["Spelldawn Team"]
homepage = "https://github.com/thurn/spelldawn"
description = "An asymmetric trading card game"
license = "Apache-2.0"
default-run = "spelldawn"
edition = "2021"
rust-version = "1.57"

[[bin]]
name = "spelldawn"
path = "src/main.rs"

[dependencies]
prost = "0.9.0"
prost-types = "0.9.0"
tokio = { version = "1.14.0", features = ["macros", "rt-multi-thread", "sync", "time"] }
tokio-stream = "0.1.8"
tonic = "0.6.1"
tonic-web = "0.2"

protos = { path = "../protos", version = "0.0.0" }