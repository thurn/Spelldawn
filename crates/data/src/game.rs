// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::card_name::CardName;
use crate::card_state;
use crate::card_state::{AbilityState, CardPosition, CardPositionKind, CardState, SortingKey};
use crate::deck::Deck;
use crate::primitives::{
    AbilityId, AbilityIndex, ActionCount, CardId, GameId, ManaValue, PointsValue, RaidId, Side,
    TurnNumber, UserId,
};
use crate::updates::{GameUpdate, UpdateTracker};
use rand::rngs::ThreadRng;
use rand::seq::IteratorRandom;
use rand::{thread_rng, Rng, RngCore};
use serde::{Deserialize, Serialize};
use std::cell::RefCell;
use std::collections::btree_map::Entry;
use std::iter;
use std::iter::{Enumerate, Map};
use std::slice::Iter;

#[derive(PartialEq, Eq, Hash, Debug, Clone, Serialize, Deserialize)]
pub struct PlayerState {
    pub id: UserId,
    pub mana: ManaValue,
    pub actions: ActionCount,
    pub score: PointsValue,
}

impl PlayerState {
    pub fn new_game(id: UserId, actions: ActionCount) -> PlayerState {
        PlayerState { id, mana: 5, actions, score: 0 }
    }
}

#[derive(PartialEq, Eq, Hash, Debug, Clone, Default, Serialize, Deserialize)]
pub struct AnimationBuffer {}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct RaidState {
    /// Unique ID for this raid
    pub raid_id: RaidId,
    /// Encounter position within this raid
    pub encounter_number: u32,
    /// Player who is next to act within this raid
    pub priority: Side,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct GameData {
    /// Current player whose turn it is
    pub turn: Side,
    /// Turn number for that player
    pub turn_number: TurnNumber,
    /// Data about an ongoing raid, if any
    pub raid: Option<RaidState>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct NewGameOptions {
    /// Whether to run in simulation mode and thus disable update tracking
    pub simulation: bool,
}

/// Stores the primary state for an ongoing game
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameState {
    /// Unique identifier for this game
    pub id: GameId,
    /// General game state
    pub data: GameData,
    /// Used to track changes to game state in order to update the client. Code which mutates the
    /// game state is responsible for appending a description of the change to `updates` via
    /// [UpdateTracker::push].
    ///
    /// A new `updates` buffer should be set for each network request to track changes in response
    /// to that request. Consequently, its value is not serialized.
    #[serde(skip)]
    pub updates: UpdateTracker,
    /// Cards for the overlord player. In general, code should use one of the helper methods below
    /// instead of accessing this directly.
    pub overlord_cards: Vec<CardState>,
    /// Cards for the champion player. In general, code should use one of the helper methods below
    /// instead of accessing this directly.
    pub champion_cards: Vec<CardState>,
    /// State for the overlord player
    pub overlord: PlayerState,
    /// State for the champion player
    pub champion: PlayerState,
    /// Next sorting key to use for card moves. Automatically updated by [Self::move_card], do not
    /// mutate this directly.
    pub next_sorting_key: SortingKey,
}

impl GameState {
    /// Creates a new game with the provided `id` and decks for both players
    pub fn new_game(
        id: GameId,
        overlord_deck: Deck,
        champion_deck: Deck,
        options: NewGameOptions,
    ) -> Self {
        Self {
            id,
            overlord_cards: Self::make_deck(&overlord_deck, Side::Overlord),
            champion_cards: Self::make_deck(&champion_deck, Side::Champion),
            overlord: PlayerState::new_game(overlord_deck.owner_id, 3 /* actions */),
            champion: PlayerState::new_game(champion_deck.owner_id, 0 /* actions */),
            data: GameData { turn: Side::Overlord, turn_number: 1, raid: None },
            updates: if options.simulation {
                UpdateTracker::default()
            } else {
                UpdateTracker { update_list: Some(vec![]) }
            },
            next_sorting_key: 1,
        }
    }

    /// Returns the identity card for the provided Side.
    ///
    /// It is an error for there to be zero or multiple cards in the `Identity` card position. If
    /// this does occur, this method will panic (in the case of zero cards) or return an arbitrary
    /// identity card (in the case of multiples).
    pub fn identity(&self, side: Side) -> &CardState {
        self.cards(side)
            .iter()
            .find(|c| c.position.kind() == CardPositionKind::Identity)
            .expect("Identity Card")
    }

    /// All Card IDs present in this game
    pub fn all_card_ids(&self) -> impl Iterator<Item = CardId> {
        (0..self.overlord_cards.len())
            .map(|index| CardId::new(Side::Overlord, index))
            .chain((0..self.champion_cards.len()).map(|index| CardId::new(Side::Champion, index)))
    }

    /// Look up [CardState] for a card
    pub fn card(&self, card_id: impl Into<CardId>) -> &CardState {
        let id = card_id.into();
        &self.cards(id.side)[id.index]
    }

    /// Mutable version of [Self::card]
    pub fn card_mut(&mut self, card_id: impl Into<CardId>) -> &mut CardState {
        let id = card_id.into();
        &mut self.cards_mut(id.side)[id.index]
    }

    /// Cards for a player
    pub fn cards(&self, side: Side) -> &Vec<CardState> {
        match side {
            Side::Overlord => &self.overlord_cards,
            Side::Champion => &self.champion_cards,
        }
    }

    /// Mutable version of [Self::cards]
    pub fn cards_mut(&mut self, side: Side) -> &mut Vec<CardState> {
        match side {
            Side::Overlord => &mut self.overlord_cards,
            Side::Champion => &mut self.champion_cards,
        }
    }

    /// State for the players in the game
    pub fn player(&self, side: Side) -> &PlayerState {
        match side {
            Side::Overlord => &self.overlord,
            Side::Champion => &self.champion,
        }
    }

    /// Mutable version of [Self::player]
    pub fn player_mut(&mut self, side: Side) -> &mut PlayerState {
        match side {
            Side::Overlord => &mut self.overlord,
            Side::Champion => &mut self.champion,
        }
    }

    /// Moves a card to a new [CardPosition].
    pub fn move_card(&mut self, card_id: impl Into<CardId>, new_position: CardPosition) {
        let key = self.next_sorting_key;
        let mut card = self.card_mut(card_id);
        card.position = new_position;
        card.sorting_key = key;
        self.next_sorting_key += 1;
    }

    /// Return a random card in the provided `position`, or None if there are no cards in that
    /// position
    pub fn random_card(&self, position: CardPosition) -> Option<CardId> {
        self.overlord_cards
            .iter()
            .chain(self.champion_cards.iter())
            .choose(&mut rand::thread_rng())
            .map(|c| c.id)
    }

    /// Cards in a player's hand
    pub fn hand(&self, side: Side) -> impl Iterator<Item = &CardState> {
        self.cards(side).iter().filter(|c| c.position.in_hand())
    }

    /// Mutable version of [Self::hand]
    pub fn hand_mut(&mut self, side: Side) -> impl Iterator<Item = &mut CardState> {
        self.cards_mut(side).iter_mut().filter(|c| c.position.in_hand())
    }

    /// Cards in a player's discard pile
    pub fn discard_pile(&self, side: Side) -> impl Iterator<Item = &CardState> {
        self.cards(side).iter().filter(|c| c.position.in_discard_pile())
    }

    /// Mutable version of [Self::discard_pile]
    pub fn discard_pile_mut(&mut self, side: Side) -> impl Iterator<Item = &mut CardState> {
        self.cards_mut(side).iter_mut().filter(|c| c.position.in_discard_pile())
    }

    /// Create card states for a deck
    fn make_deck(deck: &Deck, side: Side) -> Vec<CardState> {
        let mut result = vec![CardState::new(
            CardId::new(side, 0),
            deck.identity,
            side,
            true, /* is_identity */
        )];

        result.extend(deck.card_names().iter().enumerate().map(move |(index, name)| {
            CardState::new(CardId::new(side, index + 1), *name, side, false /* is_identity */)
        }));

        result
    }

    /// All cards in this game
    pub fn all_cards(&self) -> impl Iterator<Item = &CardState> {
        self.overlord_cards.iter().chain(self.champion_cards.iter())
    }
}
