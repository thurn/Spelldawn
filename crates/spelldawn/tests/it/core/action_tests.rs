// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use cards::test_cards::{ARTIFACT_COST, MANA_STORED, MANA_TAKEN, UNVEIL_COST};
use data::card_name::CardName;
use data::game::RaidPhase;
use data::primitives::Side;
use insta::assert_snapshot;
use protos::spelldawn::game_action::Action;
use protos::spelldawn::object_position::Position;
use protos::spelldawn::{
    card_target, CardTarget, ClientRoomLocation, DrawCardAction, GainManaAction, GameMessageType,
    LevelUpRoomAction, ObjectPositionDiscardPile, PlayCardAction, PlayerName,
};
use test_utils::summarize::Summary;
use test_utils::*;

#[test]
fn connect() {
    let mut g = new_game(Side::Overlord, Args { connect: false, ..Args::default() });
    let response = g.connect(g.user_id(), Some(g.game_id()));
    let _summary = Summary::run(&response);
    assert_snapshot!(Summary::run(&response));
}

#[test]
fn connect_to_ongoing() {
    let mut g = new_game(
        Side::Overlord,
        Args { turn_actions: 3, deck_top: Some(CardName::IceDragon), ..Args::default() },
    );
    let r1 = g.connect(g.user_id(), Some(g.game_id()));
    assert_ok(&r1);
    let r2 = g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id());
    assert_identical(vec![CardName::IceDragon], g.user.cards.hand(PlayerName::User));
    assert_ok(&r2);
    let r3 = g.connect(g.opponent_id(), Some(g.game_id()));

    assert_snapshot!(Summary::run(&r3));
}

#[test]
fn draw_card() {
    let mut g = new_game(
        Side::Overlord,
        Args { turn_actions: 3, deck_top: Some(CardName::IceDragon), ..Args::default() },
    );
    let response = g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id());
    assert_snapshot!(Summary::run(&response));

    assert_identical(vec![CardName::IceDragon], g.user.cards.hand(PlayerName::User));
    assert_eq!(vec![HIDDEN_CARD], g.opponent.cards.hand(PlayerName::Opponent));
    assert_eq!(2, g.me().actions());
    assert_eq!(2, g.opponent.other_player.actions());
}

#[test]
fn cannot_draw_card_on_opponent_turn() {
    let mut g = new_game(Side::Overlord, Args::default());
    assert_error(g.perform_action(Action::DrawCard(DrawCardAction {}), g.opponent_id()));
}

#[test]
fn cannot_draw_when_out_of_action_points() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 0, ..Args::default() });
    assert_error(g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id()));
}

#[test]
fn cannot_draw_during_raid() {
    let mut g = new_game(
        Side::Overlord,
        Args {
            raid: Some(TestRaid { phase: RaidPhase::Activation, active: false }),
            ..Args::default()
        },
    );
    assert_error(g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id()));
}

#[test]
fn play_card() {
    let mut g = new_game(Side::Champion, Args { turn_actions: 3, mana: 5, ..Args::default() });
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    let response = g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    );
    assert_snapshot!(Summary::run(&response));

    assert_eq!(2, g.me().actions());
    assert_eq!(2, g.opponent.other_player.actions());
    assert_eq!(9, g.me().mana());
    assert_eq!(9, g.opponent.other_player.mana());
    assert_identical(vec![CardName::ArcaneRecovery], g.user.cards.discard_pile(PlayerName::User));
    assert_identical(
        vec![CardName::ArcaneRecovery],
        g.opponent.cards.discard_pile(PlayerName::Opponent),
    );
}

#[test]
fn play_hidden_card() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 3, mana: 0, ..Args::default() });
    let card_id = g.add_to_hand(CardName::DungeonAnnex);
    let response = g.perform_action(
        Action::PlayCard(PlayCardAction {
            card_id: Some(card_id),
            target: Some(CardTarget {
                card_target: Some(card_target::CardTarget::RoomId(CLIENT_ROOM_ID.into())),
            }),
        }),
        g.user_id(),
    );
    assert_snapshot!(Summary::run(&response));

    assert_eq!(2, g.me().actions());
    assert_eq!(2, g.opponent.other_player.actions());
    assert_eq!(0, g.me().mana());
    assert_eq!(0, g.opponent.other_player.mana());
    assert_identical(
        vec![CardName::DungeonAnnex],
        g.user.cards.room_cards(ROOM_ID, ClientRoomLocation::Back),
    );
    assert_eq!(vec![HIDDEN_CARD], g.opponent.cards.room_cards(ROOM_ID, ClientRoomLocation::Back));
}

#[test]
fn cannot_play_card_on_opponent_turn() {
    let mut g = new_game(Side::Overlord, Args::default());
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    assert_error(g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    ));
}

#[test]
fn cannot_play_card_when_out_of_action_points() {
    let mut g = new_game(Side::Champion, Args { turn_actions: 0, ..Args::default() });
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    assert_error(g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    ));
}

#[test]
fn cannot_play_card_during_raid() {
    let mut g = new_game(
        Side::Champion,
        Args {
            raid: Some(TestRaid { phase: RaidPhase::Activation, active: false }),
            ..Args::default()
        },
    );
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    assert_error(g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    ));
}

#[test]
fn gain_mana() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 3, mana: 5, ..Args::default() });
    let response = g.perform_action(Action::GainMana(GainManaAction {}), g.user_id());

    assert_eq!(2, g.me().actions());
    assert_eq!(2, g.opponent.other_player.actions());
    assert_eq!(6, g.me().mana());
    assert_eq!(6, g.opponent.other_player.mana());

    assert_snapshot!(Summary::run(&response));
}

#[test]
fn cannot_gain_mana_on_opponent_turn() {
    let mut g = new_game(Side::Overlord, Args::default());
    assert_error(g.perform_action(Action::GainMana(GainManaAction {}), g.opponent_id()));
}

#[test]
fn cannot_gain_mana_when_out_of_action_points() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 0, ..Args::default() });
    assert_error(g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()));
}

#[test]
fn cannot_gain_mana_during_raid() {
    let mut g = new_game(
        Side::Overlord,
        Args {
            raid: Some(TestRaid { phase: RaidPhase::Activation, active: false }),
            ..Args::default()
        },
    );
    assert_error(g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()));
}

#[test]
fn level_up_room() {
    let mut g = new_game(Side::Overlord, Args { mana: 10, ..Args::default() });
    g.play_from_hand(CardName::TestScheme31);
    let response = g.perform_action(
        Action::LevelUpRoom(LevelUpRoomAction { room_id: CLIENT_ROOM_ID.into() }),
        g.user_id(),
    );

    assert_snapshot!(Summary::run(&response));
    assert_eq!(g.user.this_player.mana(), 9);
    assert_eq!(g.opponent.other_player.mana(), 9);
}

#[test]
fn score_overlord_card() {
    let mut g = new_game(Side::Overlord, Args { mana: 10, turn_actions: 5, ..Args::default() });
    let scheme_id = g.play_from_hand(CardName::TestScheme31);
    let level_up = Action::LevelUpRoom(LevelUpRoomAction { room_id: CLIENT_ROOM_ID.into() });
    g.perform(level_up.clone(), g.user_id());
    g.perform(level_up.clone(), g.user_id());
    let response = g.perform_action(level_up, g.user_id());

    assert_snapshot!(Summary::run(&response));
    assert!(g.opponent.cards.get(scheme_id).revealed_to_me());
    assert_eq!(g.user.this_player.mana(), 7);
    assert_eq!(g.opponent.other_player.mana(), 7);
    assert_eq!(g.user.this_player.score(), 1);
    assert_eq!(g.opponent.other_player.score(), 1);
}

#[test]
fn overlord_win_game() {
    let mut g =
        new_game(Side::Overlord, Args { mana: 10, score: 6, turn_actions: 5, ..Args::default() });
    g.play_from_hand(CardName::TestScheme31);
    let level_up = Action::LevelUpRoom(LevelUpRoomAction { room_id: CLIENT_ROOM_ID.into() });
    g.perform(level_up.clone(), g.user_id());
    g.perform(level_up.clone(), g.user_id());
    let response = g.perform_action(level_up, g.user_id());

    assert_snapshot!(Summary::run(&response));
    assert_eq!(g.user.data.last_message(), GameMessageType::Victory);
    assert_eq!(g.opponent.data.last_message(), GameMessageType::Defeat);
}

#[test]
fn switch_turn() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 3, mana: 5, ..Args::default() });
    g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()).unwrap();
    g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()).unwrap();
    let response = g.perform_action(Action::GainMana(GainManaAction {}), g.user_id());
    assert_snapshot!(Summary::run(&response));

    assert_eq!(8, g.me().mana());
    assert_eq!(8, g.opponent.other_player.mana());
    assert_eq!(0, g.me().actions());
    assert_eq!(0, g.opponent.other_player.actions());
    assert_eq!(3, g.user.other_player.actions());
    assert_eq!(3, g.opponent.this_player.actions());
    assert_eq!(g.user.data.priority(), PlayerName::Opponent);
    assert_eq!(g.opponent.data.priority(), PlayerName::User);
}

#[test]
fn activate_ability() {
    let mut g = new_game(Side::Champion, Args { turn_actions: 3, ..Args::default() });
    g.play_from_hand(CardName::TestActivatedAbilityTakeMana);
    let ability_card_id = g
        .user
        .cards
        .cards_in_hand(PlayerName::User)
        .find(|c| c.id().ability_id.is_some())
        .expect("ability card")
        .id();

    assert_eq!(STARTING_MANA - ARTIFACT_COST, g.me().mana());
    assert_eq!(2, g.me().actions());

    let response = g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(ability_card_id), target: None }),
        g.user_id(),
    );

    assert_snapshot!(Summary::run(&response));
    assert_eq!(STARTING_MANA - ARTIFACT_COST + MANA_TAKEN, g.me().mana());
    assert_eq!(1, g.me().actions());
}

#[test]
fn activate_ability_take_all_mana() {
    let mut g = new_game(Side::Champion, Args { turn_actions: 3, ..Args::default() });
    let id = g.play_from_hand(CardName::TestActivatedAbilityTakeMana);
    let ability_card_id = g
        .user
        .cards
        .cards_in_hand(PlayerName::User)
        .find(|c| c.id().ability_id.is_some())
        .expect("ability card")
        .id();

    let mut taken = 0;
    while taken < MANA_STORED {
        g.perform(
            Action::PlayCard(PlayCardAction { card_id: Some(ability_card_id), target: None }),
            g.user_id(),
        );
        taken += MANA_TAKEN;

        draw_cards_until_turn_over(&mut g, Side::Champion);
        assert!(g.dusk());
        draw_cards_until_turn_over(&mut g, Side::Overlord);
        assert!(g.dawn());
    }

    assert_eq!(STARTING_MANA - ARTIFACT_COST + MANA_STORED, g.user.this_player.mana());
    assert_eq!(STARTING_MANA - ARTIFACT_COST + MANA_STORED, g.opponent.other_player.mana());
    assert_eq!(
        Position::DiscardPile(ObjectPositionDiscardPile { owner: PlayerName::User.into() }),
        g.user.cards.get(id).position()
    );
    assert_eq!(
        Position::DiscardPile(ObjectPositionDiscardPile { owner: PlayerName::Opponent.into() }),
        g.opponent.cards.get(id).position()
    );
}

#[test]
fn triggered_ability() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 1, ..Args::default() });
    g.play_from_hand(CardName::TestTriggeredAbilityTakeManaAtDusk);
    assert!(g.dawn());
    assert_eq!(STARTING_MANA, g.user.this_player.mana());
    g.perform(Action::GainMana(GainManaAction {}), g.opponent_id());
    g.perform(Action::GainMana(GainManaAction {}), g.opponent_id());
    let response = g.perform_action(Action::GainMana(GainManaAction {}), g.opponent_id());
    assert!(g.dusk());
    assert_snapshot!(Summary::run(&response));
    assert_eq!(STARTING_MANA - UNVEIL_COST + MANA_TAKEN, g.user.this_player.mana());
    assert_eq!(STARTING_MANA - UNVEIL_COST + MANA_TAKEN, g.opponent.other_player.mana());
}

#[test]
fn triggered_ability_cannot_unveil() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 1, mana: 0, ..Args::default() });
    g.play_from_hand(CardName::TestTriggeredAbilityTakeManaAtDusk);
    assert!(g.dawn());
    assert_eq!(0, g.user.this_player.mana());
    gain_mana_until_turn_over(&mut g, Side::Champion);
    assert!(g.dusk());
    assert_eq!(0, g.user.this_player.mana());
    assert_eq!(0, g.opponent.other_player.mana());
}

#[test]
fn triggered_ability_take_all_mana() {
    let mut g = new_game(Side::Overlord, Args { turn_actions: 1, ..Args::default() });
    let id = g.play_from_hand(CardName::TestTriggeredAbilityTakeManaAtDusk);
    let mut taken = 0;
    while taken < MANA_STORED {
        assert!(g.dawn());
        gain_mana_until_turn_over(&mut g, Side::Champion);
        assert!(g.dusk());
        taken += MANA_TAKEN;
        draw_cards_until_turn_over(&mut g, Side::Overlord);
    }

    assert_eq!(STARTING_MANA - UNVEIL_COST + MANA_STORED, g.user.this_player.mana());
    assert_eq!(STARTING_MANA - UNVEIL_COST + MANA_STORED, g.opponent.other_player.mana());
    assert_eq!(
        Position::DiscardPile(ObjectPositionDiscardPile { owner: PlayerName::User.into() }),
        g.user.cards.get(id).position()
    );
    assert_eq!(
        Position::DiscardPile(ObjectPositionDiscardPile { owner: PlayerName::Opponent.into() }),
        g.opponent.cards.get(id).position()
    );
}
