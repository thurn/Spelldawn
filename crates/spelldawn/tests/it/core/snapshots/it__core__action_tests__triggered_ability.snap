---
source: crates/spelldawn/tests/it/core/action_tests.rs
assertion_line: 371
expression: "Summary::run(&response)"
---

command_list: 
    CreateOrUpdateCard: 
        card: 
            card_id: O1
            revealed_to_viewer: false
            is_face_up: false
            card_icons: 
                top_left_icon: 
                top_right_icon: 
                bottom_left_icon: 
                bottom_right_icon: 
                arena_icon: 
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
        create_position: 
            sorting_key: 4294967295
            position: ObjectPositionDeck { owner: Opponent }
        create_animation: Unspecified
    MoveGameObjects: 
        id: O1
        position: 
            sorting_key: 6
            position: ObjectPositionHand { owner: Opponent }
    UpdateGameView: 
        user: 
            side: Champion
            player_info: 
            valid_rooms_to_visit: 
            mana: 1002
            action_tracker: 0
            can_take_action: false
        opponent: 
            side: Overlord
            mana: 998
            action_tracker: 3
            can_take_action: true
        raid_active: false
    CreateOrUpdateCard: 
        card: 
            card_id: C1
            revealed_to_viewer: true
            is_face_up: false
            owning_player: User
            revealed_card: 
                targeting: 
                    can_play: false
        create_position: 
            sorting_key: 4
            position: ObjectPositionHand { owner: User }
        create_animation: Unspecified
    MoveGameObjects: 
        id: O45
        position: 
            sorting_key: 0
            position: ObjectPositionStaging
    CreateOrUpdateCard: 
        card: 
            card_id: O45
            revealed_to_viewer: true
            is_face_up: true
            card_icons: 
                top_left_icon: "3"
                top_right_icon: 
                bottom_left_icon: 
                arena_icon: "8"
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Triggered Ability Take Mana At Dusk"
                rules_text: "<RulesText>"
                targeting: 
                valid_rooms: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionRoom { room_id: Unspecified, room_location: Back }
        create_position: 
            sorting_key: 2
            position: ObjectPositionRoom { room_id: RoomA, room_location: Back }
        create_animation: Unspecified
    Delay: 1500
    MoveGameObjects: 
        id: O45
        position: 
            sorting_key: 2
            position: ObjectPositionRoom { room_id: RoomA, room_location: Back }
    CreateOrUpdateCard: 
        card: 
            card_id: O45[1]
            revealed_to_viewer: true
            is_face_up: false
            card_icons: 
                top_left_icon: 
            owning_player: Opponent
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Use Test Triggered Ability Take Mana At Dusk"
                rules_text: "<RulesText>"
                on_release_position: 
                    sorting_key: 2
                    position: ObjectPositionIntoCard { card_id: Some(CardIdentifier { side: Overlord, index: 45, ability_id: None }) }
        create_position: 
            sorting_key: 2
            position: ObjectPositionStaging
        create_animation: FromParentCard
    Delay: 1500
    MoveGameObjects: 
        id: O45[1]
        position: 
            sorting_key: 0
            position: ObjectPositionIntoCard { card_id: Some(CardIdentifier { side: Overlord, index: 45, ability_id: None }) }
    DestroyCard: 
        card_id: O45[1]
    DisplayGameMessage: Dusk
channel_response: 
    CreateOrUpdateCard: 
        card: 
            card_id: O1
            revealed_to_viewer: true
            is_face_up: false
            card_icons: 
                top_left_icon: "1"
                top_right_icon: 
                bottom_left_icon: 
                arena_icon: 
            arena_frame: "<SpriteAddress>"
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Overlord Spell"
                rules_text: "<RulesText>"
                targeting: 
                    can_play: true
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
        create_animation: DrawCard
    MoveGameObjects: 
        id: O1
        position: 
            sorting_key: 6
            position: ObjectPositionHand { owner: User }
    UpdateGameView: 
        user: 
            side: Overlord
            mana: 998
            action_tracker: 3
            can_take_action: true
        opponent: 
            side: Champion
            player_info: 
            valid_rooms_to_visit: 
            mana: 1002
            action_tracker: 0
            can_take_action: false
        raid_active: false
    CreateOrUpdateCard: 
        card: 
            card_id: O45
            revealed_to_viewer: true
            is_face_up: true
            card_icons: 
                arena_icon: "8"
            owning_player: User
        create_position: 
            sorting_key: 2
            position: ObjectPositionRoom { room_id: RoomA, room_location: Back }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: O45[1]
            revealed_to_viewer: true
            is_face_up: false
            card_icons: 
                top_left_icon: 
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Use Test Triggered Ability Take Mana At Dusk"
                rules_text: "<RulesText>"
                on_release_position: 
                    sorting_key: 2
                    position: ObjectPositionIntoCard { card_id: Some(CardIdentifier { side: Overlord, index: 45, ability_id: None }) }
        create_position: 
            sorting_key: 2
            position: ObjectPositionStaging
        create_animation: FromParentCard
    Delay: 1500
    MoveGameObjects: 
        id: O45[1]
        position: 
            sorting_key: 0
            position: ObjectPositionIntoCard { card_id: Some(CardIdentifier { side: Overlord, index: 45, ability_id: None }) }
    DestroyCard: 
        card_id: O45[1]
    DisplayGameMessage: Dusk
